engine/runtime-compiler/src/main/java/org/enso/compiler/phase/ImportResolverAlgorithm.java

```diff
@@ -0,0 +1,201 @@
+package org.enso.compiler.phase;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import org.enso.compiler.core.CompilerError;
+import org.enso.editions.LibraryName;
+import scala.Tuple2;
+
+public abstract class ImportResolverAlgorithm<
+    Result, Module, Import, Export, ResolvedType, ResolvedModule> {
+  protected ImportResolverAlgorithm() {}
+
+  protected abstract String nameForImport(Import imp);
+
+  protected abstract List<String> partsForImport(Import imp);
+
+  protected abstract String nameForExport(Export ex);
+
+  protected abstract String nameForType(ResolvedType e);
+
+  protected abstract java.util.List<Export> exportsFor(Module module, String impName);
+
+  protected abstract boolean isAll(Export ex);
+
+  /**
+   * @return {@code null} or list of named imports
+   */
+  protected abstract java.util.List<String> onlyNames(Export ex);
+
+  /**
+   * @return {@code null} or list of named imports
+   */
+  protected abstract java.util.List<String> hiddenNames(Export ex);
```

Comment:
I don't like that one has to check if those methods return null. We could have a more predictable API by returning empty list and at the usage site use `flatMap` instead of filtering non-null values.

Reply:
The code remained unchanged compared to Scala version. [hiddenNames](https://github.com/enso-org/enso/pull/9539/files#diff-a3b518b6d9b41b5aec1a3c6c3a2153e38cfd594cd1c11304c094ececdd4cc6cbL213) is an `Option` in Scala.

Java way to express option is to return `null`, callers have an option of `Optional.of(hiddenNames)` to convert it to an `Optional` when they find `if (hiddenNames != null` too old fashioned.